adr/ADR-33.md

```diff
@@ -0,0 +1,236 @@
+
+Services should respond to:
+
+- All service requests
+- All service requests that match their `name`
+- All services requests that match their `name` and `id`
+
+### INFO
+
+Returns a JSON having the following structure:
+
+```typescript
+{
+    /**
+     * The kind of the service reporting the status
+     */
+    name: string,
+    /**
+     * The unique ID of the service reporting the status
+     */
+    id: string,
+    /**
+     * Description for the service
+     */
+    description: string,
+    /**
+     * Version of the service
+     */
+    version: string,
+    /**
+     * Subject where the service can be invoked
+     */
+    subject: string
+}
+```
+
+All the fields above map 1-1 to the metadata provided when the service was
+created. Note that `subject` is the subject that the service is listening for
+requests on.
+
+### PING
+
+Is an alias for `INFO`. On receiving a `PING` request, the service should
+respond by sending an info response.
+
+### SCHEMA
+
+Returns a JSON having the following structure. Note that the `schema` struct is
+only returned if the `schema` was specified when created.
+
+```typescript
+{
+    /**
+     * The kind of the service reporting the status
+     */
+    name: string,
+    /**
+     * The unique ID of the service reporting the status
+     */
+    id: string,
+    /**
+     * Version of the service
+     */
+    version: string,
+    /**
+     * The schema specified when the service was created
+     */
+    schema?: {
+        /**
+         * A string or URL
+         */
+        request: string,
+        /**
+         * A string or URL
+         */
+        response: string
+    };
+}
+```
+
+### STATUS
+
+The STATUS request can have an optional JSON payload `{ internal: boolean }`. If
+specified and `true` the service should report all its internal endpoints, thus
+the reson the stats returns an array of EndpointStats:
+
+```typescript
+{
+  name: string;
+  id: string;
+  version: string;
+  stats: EndpointStats[];
+}
+```
+
+EndpointStats looks like:
+
+```typescript
+{
+    /**
+     * Name of the endpoint
+     */
+    name: string;
+    /**
+     * The number of requests received by the endpoint
+     */
+    num_requests: number;
+    /**
+     * Number of errors that the endpoint has raised
+     */
+    num_errors: number;
+    /**
+     * If set, the last error triggered by the endpoint
+     */
+    last_error ? : Error;
+    /**
+     * A field that can be customized with any data as returned by status handler see {@link ServiceConfig}
+     */
+    data ? : unknown;
+    /**
+     * Total latency for the service (in nanos)
+     */
+    total_latency: number;
+    /**
+     * Average latency is the total latency divided by the num_requests (in nanos)
+     */
+    average_latency: number;
+}
+```
+
+## Error Handling
+
+Services may communicate request errors back to the client as they see fit, but
+to help standardization they also must include the header: `Nats-Service-Error`,
```

Comment:
Don't we want to split this into 2 headers (`Nats-Service-Error` and `Nats-Service-Error-Code`)? That would eliminate the need of using helper functions to parse the header. We could add a helper function to set the headers on error though:
```go
func SetError(headers Headers, errCode int, err error)
```

Reply:
Yes we do - it would simplify parsing.

Reply:
done